### PR TITLE
feat(dashboard): Filters for activity feed

### PR DESCRIPTION
**Pull request by scopsy**
### What changed? Why was the change needed?
- New reusable multi-faced filter component
- Applied filters on the activity feed page
- 

### Screenshots
https://www.loom.com/share/f818c89fcc264ce3b5d77c403f78898e
CleanShot 2024-12-10 at 13 42 17@2x

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: apps/dashboard/src/api/api.client.ts [hunk 1]
```diff
@@ -35,6 +36,7 @@ const request = async <T>(
         ...(environment && { 'Novu-Environment-Id': environment._id }),
         ...headers,
       },
+      signal,
```
Comment (scopsy) [inline apps/dashboard/src/api/api.client.ts:39]: Added support for aborting requests

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/base-filter-content.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  showNavigationFooter?: boolean;
+  showEnterIcon?: boolean;
+  children?: React.ReactNode;
+}
+
+export function BaseFilterContent({
+  inputRef,
+  title,
+  onClear,
+  size,
```
Comment (scopsy) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/base-filter-content.tsx:26]: This folder is a new Form MultiFacted component that supports multiple field types

CleanShot 2024-12-10 at 13 55 45@2x

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/base-filter-content.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          showEnterIcon={showEnterIcon}
+        />
+      )}
+
+      {children}
+
+      {showNavigationFooter && (
+        <div className="flex justify-between rounded-b-md border-t border-neutral-100 bg-white p-1">
+          <div className="flex items-center gap-0.5">
+            <div className="pointer-events-none shrink-0 rounded-[6px] border border-neutral-200 bg-white p-1 shadow-[0px_0px_0px_1px_rgba(14,18,27,0.02)_inset,_0px_1px_4px_0px_rgba(14,18,27,0.12)]">
```
Comment (scopsy) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/base-filter-content.tsx:59]: Our shadows currently not synchronized with Figma. Will schedule a session with naveen to fix it across the board. Added it manually here for now

File: apps/dashboard/src/components/activity/activity-table.tsx [hunk 4]
```diff
@@ -64,61 +76,86 @@ export function ActivityTable({ selectedActivityId, onActivitySelect }: Activity
...
+    <AnimatePresence mode="wait" initial={false}>
+      {!isLoading && activities.length === 0 ? (
+        <motion.div
+          key="empty-state"
+          initial={{ opacity: 0 }}
+          animate={{ opacity: 1 }}
+          exit={{ opacity: 0 }}
+          transition={{ duration: 0.2 }}
+          className="flex h-full w-full items-center justify-center"
+        >
```
Comment (scopsy) [inline apps/dashboard/src/components/activity/activity-table.tsx:69]: Prevents flicker between loading and empty state transitions

File: apps/api/src/app/notifications/notification.controller.ts [hunk 5]
```diff
@@ -77,6 +77,8 @@ export class NotificationsController {
         search: query.search,
         subscriberIds: subscribersQuery,
         transactionId: query.transactionId,
+        startDate: query.startDate,
```
Comment (SokratisVidros) [inline apps/api/src/app/notifications/notification.controller.ts]:
```suggestion
        after: query.after,
```

File: apps/api/src/app/notifications/notification.controller.ts [hunk 6]
```diff
@@ -77,6 +77,8 @@ export class NotificationsController {
...
+        endDate: query.endDate,
```
Comment (SokratisVidros) [inline apps/api/src/app/notifications/notification.controller.ts]:
```suggestion
        before: query.before,
```

File: apps/api/src/app/notifications/usecases/get-activity-feed/get-activity-feed.command.ts [hunk 7]
```diff
@@ -33,4 +33,12 @@ export class GetActivityFeedCommand extends EnvironmentWithUserCommand {
   @IsOptional()
   @IsString()
   transactionId?: string;
+
+  @IsOptional()
+  @IsString()
+  startDate?: string;
```
Comment (SokratisVidros) [inline apps/api/src/app/notifications/usecases/get-activity-feed/get-activity-feed.command.ts]: Ditto

File: apps/api/src/app/notifications/usecases/get-activity-feed/get-activity-feed.command.ts [hunk 8]
```diff
@@ -33,4 +33,12 @@ export class GetActivityFeedCommand extends EnvironmentWithUserCommand {
...
+
+  @IsOptional()
+  @IsString()
+  endDate?: string;
```
Comment (SokratisVidros) [inline apps/api/src/app/notifications/usecases/get-activity-feed/get-activity-feed.command.ts]: Ditto

File: apps/dashboard/src/components/activity/activity-empty-state.tsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            initial={{ opacity: 0, y: 5 }}
+            animate={{ opacity: 1, y: 0 }}
+            transition={{
+              duration: 0.2,
+              delay: 0.25,
+            }}
+            className="flex flex-col items-center gap-1 text-center"
+          >
+            <h2 className="text-foreground-900 text-lg font-medium">
+              {emptySearchResults ? 'No activity match that filter' : 'No activity in the past 30 days'}
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/activity/activity-empty-state.tsx:71]: If not mistaken, the past 30 days should either be past 30 or past 90, depending on the plan. Or we can just rephrase it to `No activity in the selected interval`.

File: apps/dashboard/src/components/activity/activity-empty-state.tsx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+            </h2>
+            <p className="text-foreground-600 max-w-md text-sm font-normal">
+              {emptySearchResults
+                ? 'Try changing your filter to see more activity or trigger notifications that match your search criteria.'
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/activity/activity-empty-state.tsx]:
```suggestion
                ? 'Change your search criteria.'
```

File: apps/dashboard/src/components/activity/activity-empty-state.tsx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                : "Your activity feed is empty. Once events start appearing, you'll be able to track notifications, troubleshoot issues, and view delivery details."}
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/activity/activity-empty-state.tsx]:
```suggestion
                : "Your activity feed is empty. Once you trigger your first workflow, you can monitor notifications and view delivery details."}
```

File: apps/dashboard/src/components/activity/activity-filters.tsx [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+import { FacetedFormFilter } from '../primitives/form/faceted-filter/facated-form-filter';
+import { CalendarIcon } from 'lucide-react';
+
+export interface IActivityFilters {
+  onFiltersChange: (filters: IActivityFiltersData) => void;
+  initialValues: IActivityFiltersData;
+  onReset?: () => void;
+}
+
+export interface IActivityFiltersData {
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/activity/activity-filters.tsx]:
```suggestion
export type = IActivityFiltersData {
```

File: apps/dashboard/src/components/activity/activity-filters.tsx [hunk 13]
```diff
@@ -0,0 +1,182 @@
+import { useEffect, useMemo } from 'react';
+import { ChannelTypeEnum } from '@novu/shared';
+import { useFetchWorkflows } from '../../hooks/use-fetch-workflows';
+import { useForm } from 'react-hook-form';
+import { Form, FormItem, FormField } from '../primitives/form/form';
+import { Button } from '../primitives/button';
+import { FacetedFormFilter } from '../primitives/form/faceted-filter/facated-form-filter';
+import { CalendarIcon } from 'lucide-react';
+
+export interface IActivityFilters {
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/activity/activity-filters.tsx]:
```suggestion
export type = IActivityFilters {
```

File: apps/dashboard/src/components/activity/activity-filters.tsx [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+  { value: ChannelTypeEnum.IN_APP, label: 'In-App' },
+  { value: ChannelTypeEnum.PUSH, label: 'Push' },
+];
+
+export const defaultActivityFilters: IActivityFiltersData = {
+  dateRange: '30d',
+  channels: [],
+  workflows: [],
+  transactionId: '',
+  subscriberId: '',
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/activity/activity-filters.tsx:43]: What about after and before dates?

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/base-filter-content.tsx [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+          inputRef={inputRef}
+          value={searchValue}
+          onChange={onSearchChange}
+          placeholder={searchPlaceholder}
+          size={size}
+          showEnterIcon={showEnterIcon}
+        />
+      )}
+
+      {children}
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/base-filter-content.tsx]: I suggest wrapping children in a scrollable container so that the component has a max height when the list items are more than 8 or 10.

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/clear-button.tsx [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+  onClick: () => void;
+  size: SizeType;
+  label?: string;
+  className?: string;
+  separatorClassName?: string;
+}
+
+export function ClearButton({ onClick, size, label = 'Clear filter', className }: ClearButtonProps) {
+  return (
+    <>
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/clear-button.tsx]: Please remove the fragment.

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/filter-input.tsx [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  value: string;
+  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  placeholder?: string;
+  size: SizeType;
+  showEnterIcon?: boolean;
+}
+
+export function FilterInput({ inputRef, value, onChange, placeholder, size, showEnterIcon = false }: FilterInputProps) {
+  return (
+    <div className="px-2 py-1">
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/filter-input.tsx]: Can this div be combine with the one that follows?

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/multi-filter-content.tsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
+import { Check } from 'lucide-react';
+import { cn } from '../../../../../utils/ui';
+import { FilterOption, SizeType } from '../types';
+import { BaseFilterContent } from './base-filter-content';
+import { useKeyboardNavigation } from '../hooks/use-keyboard-navigation';
+
+interface MultiFilterContentProps {
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/multi-filter-content.tsx]: props

File: apps/dashboard/src/components/primitives/form/faceted-filter/components/single-filter-content.tsx [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+      onClear={onClear}
+      size={size}
+      hideSearch={hideSearch}
+      hideClear={hideClear}
+      searchValue={searchQuery}
+      onSearchChange={handleSearchChange}
+      searchPlaceholder={`Search ${title}...`}
+      showNavigationFooter={true}
+    >
+      <div className={cn('flex flex-col gap-1 p-1')}>
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/form/faceted-filter/components/single-filter-content.tsx]: The className props can be passes as a className prop to the RadioGroup

File: apps/dashboard/src/components/primitives/form/faceted-filter/facated-form-filter.tsx [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+    } else {
+      onSelect?.([]);
+    }
+    setSearchQuery('');
+  };
+
+  const renderTriggerContent = () => {
+    if (type === 'text' && currentValue) {
+      return (
+        <>
```
Comment (SokratisVidros) [inline apps/dashboard/src/components/primitives/form/faceted-filter/facated-form-filter.tsx]:
```suggestion
```

File: apps/dashboard/src/api/api.client.ts (hunk 1 above)
Comment (SokratisVidros) [reply apps/dashboard/src/api/api.client.ts:39]: Are we using the signal in this PR?

**Review by SokratisVidros — CHANGES_REQUESTED** [review]
@LetItRock @desiprisg @BiswaViraj as you can see in this PR we need single and multi select dropdown. For now they are introduced as part of the faceted filters. The functionality is kind of similar to the Tags component.

Please coordinate with Dima on how to archive reusability by introducing the necessary primitives. Let me know about the necessary actions so that we can add it as a task to linear.

There is also [this component](https://shadcn-multi-select-component.vercel.app/) that looks it meets our requirements.

**Review by LetItRock — CHANGES_REQUESTED** [review]
@scopsy something is broken here
Screenshot 2024-12-12 at 10 12 34

Also we don't handle properly the pagination. It allows me to go to the next empty page and at the result I see this:
Screenshot 2024-12-12 at 10 16 42

The workflows list can be huge and should be scrollable:
Screenshot 2024-12-12 at 10 21 56

Missing the `Chat` in the filter:
Screenshot 2024-12-12 at 10 22 16

The table is loading when picking the item from the list:

https://github.com/user-attachments/assets/a937c71e-7f3b-4608-89d4-f91e69c1f156

**Comment by LetItRock** [comment]
No reset button on channels filter, same about the workflows
Screenshot 2024-12-12 at 10 29 24

File: apps/dashboard/src/components/activity/activity-filters.tsx (hunk 14 above)
Comment (scopsy) [reply apps/dashboard/src/components/activity/activity-filters.tsx:43]: They are specified as dateRange in the UI, before and after is a backend implementation

File: apps/dashboard/src/components/activity/activity-empty-state.tsx (hunk 9 above)
Comment (scopsy) [reply apps/dashboard/src/components/activity/activity-empty-state.tsx:71]: Any change to the default interval will result in the `emptySearchResults` this message only target the default state result. And mostly targeting new signups
